core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -145,61 +203,67 @@ private ObjectNode generate(MethodInfo methodInfo) {
      * and add required fields to the {@link ArrayNode}.
      *
      * @param field field to generate schema for
-     * @param visited map of visited types and their paths
      * @param path current path in tree traversal of fields
      * @param parent the parent to add schema properties
      * @param required the array node to add required field names, if parent doesn't support, it is null.
      */
-    private void generateField(FieldInfo field, Map<TypeSignature, String> visited, String path,
-                               ObjectNode parent,
-                               @Nullable ArrayNode required) {
+    private void generateField(FieldInfo field, String path, ObjectNode parent, @Nullable ArrayNode required) {
         final ObjectNode fieldNode = mapper.createObjectNode();
         final TypeSignature fieldTypeSignature = field.typeSignature();
+        final String schemaType = getSchemaType(fieldTypeSignature);
+        final String currentPath;
 
-        fieldNode.put("description", field.descriptionInfo().docString());
+        if (field.name().isEmpty()) {
+            currentPath = path;
+        } else {
+            currentPath = path + '/' + field.name();
+        }
+
+        final String description = field.descriptionInfo().docString();
+        if (!description.isEmpty()) {
+            fieldNode.put("description", description);
+        }
 
         // Fill required fields for the current object.
         if (required != null && field.requirement() == FieldRequirement.REQUIRED) {
             required.add(field.name());
         }
 
+        final boolean shouldFlatten = useFlatSchema &&
+                                      MEMORIZED_JSON_TYPES.contains(schemaType) &&
+                                      currentPath.chars().filter(c -> c == '/').count() > 1;
+        if (shouldFlatten) {
+            // If field is already visited, add a reference to the field instead of iterating its children.
+            if (!visited.containsKey(fieldTypeSignature)) {
+                final ObjectNode defsFieldNode = mapper.createObjectNode();
+                final String currentDefsPath = DEFS_PATH + '/' + fieldTypeSignature.signature();
+                visited.put(fieldTypeSignature, currentDefsPath);
+
+                generateTypeFields(defsFieldNode, field, schemaType, currentDefsPath);
+
+                defsNode.set(fieldTypeSignature.signature(), defsFieldNode);
```

Comment:
Based on the [generated gist](https://gist.github.com/Dogacel/bfa94713eb3202d4a1f02f057ad1a066#file-schema-defs-json-L113-L130), `repeated` and `map` are defined as `defs` which is awkward. What do you think of embedding `repeated` and `map` and referencing their elements if they are non-primitive types?

```diff
- "strings" : {
-  "description" : "",
-  "$ref" : "#/$defs/repeated<string>"
- }
+ "strings" : {
+  "description" : "",
+  "type": "array",
+  "items" : {
+    "description" : "",
+    "type" : "string"
+  }
+ }
```
Otherwise, is `$defs/repeated<string>` necessary to create an OpenAPI specification?

Reply:
Let me check, I think what you suggested makes sense.

Reply:
@Dogacel Did you have a chance to check it?

Reply:
Coming back to this after another month (I don't understand how time flies by).

Concern is valid, I agree.

I have updated the code to include a new method called `canReferenceAsDef`. This method returns true if the type is another object but not array or map. So we are not creating weird references to primitives or at least trivial types.

Please take another look 😄